src/history/HistoryArchive.cpp

```diff
@@ -259,6 +262,83 @@ HistoryArchiveState::allBuckets() const
     return std::vector<std::string>(buckets.begin(), buckets.end());
 }
 
+bool
+HistoryArchiveState::containsValidBuckets(Application& app) const
+{
+    // This function assumes presence of required buckets to verify state
+    // Level 0 future buckets are always clear
+    assert(currentBuckets[0].next.isClear());
+
+    for (uint32_t i = 1; i < BucketList::kNumLevels; i++)
+    {
+        auto& level = currentBuckets[i];
+        auto& prev = currentBuckets[i - 1];
+        Hash const emptyHash;
+
+        auto snap =
+            app.getBucketManager().getBucketByHash(hexToBin256(prev.snap));
+        assert(snap);
+        if (snap->getHash() == emptyHash)
+        {
+            continue;
+        }
+        else if (Bucket::getBucketVersion(snap) >=
+                 Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED)
+        {
+            if (!level.next.isClear())
+            {
+                CLOG(INFO, "History") << "Invalid HAS: future must be cleared ";
```

Comment:
`ERROR` log level (and possibly throw?)

Reply:
agreed that it should be `ERROR`, but i don't think throwing is right here: the node itself is not in a corrupted state, but the archive it downloaded from is.

Reply:
👍